R/define_data.R

```diff
@@ -216,6 +216,135 @@ defDataAdd <- function(dtDefs = NULL,
+#' @param id A string indicating the field name for the unique record identifier
+#' @return A data.table named dtName that is an updated data definitions table
+#' @seealso [distributions]
+#' @details The possible data distributions are: `r paste0(.getDists(),collapse = ", ")`.
+#'
+#' @examples
+#' def <- defRepeat(nvars = 4, prefix = "g", formula = "1/3;1/3;1/3", 
+#'    variance = 0, dist = "categorical")
+#' def <- defData(def, varname = "a", formula = "1;1", dist = "trtAssign")
+#' def <- defRepeat(def, 8, "b", formula = "5 + a", variance = 3, dist = "normal")
+#' def <- defData(def, "y", formula = "0.10", dist = "binary")
+#' 
+#' def
+#' @export
+#' @concept define_data
+defRepeat <- function(dtDefs = NULL,
+                    nvars,
+                    prefix,
+                    formula,
+                    variance = 0,
+                    dist = "normal",
+                    link = "identity",
+                    id = "id") {
+  
+  assertNotMissing(nvars = missing(nvars), 
+                   prefix = missing(prefix), 
+                   formula = missing(formula))
+  
+  varnames <- paste0(prefix, 1 : nvars)
+  
+  if (is.null(dtDefs)) {
+    
+    defNew <- defData(varname = varnames[1], formula = formula,
+                 variance = variance, dist=dist, link = link, id = id)
+    
+    for (i in (2:nvars) ) {
+      defNew <- defData(defNew, varname = varnames[i], 
+                  formula = formula, variance = variance, 
+                  dist=dist, link = link, id = id)
+    }
+    
+  } else {
+    
+    defNew <- data.table::copy(dtDefs)
+    
+    for (i in 1:nvars) {
+      defNew <- defData(defNew, varname = varnames[i], 
+                  formula = formula, variance = variance, 
+                  dist=dist, link = link, id = id)
+    }
+  }
+  
+  return(defNew[])
+}
+
+#' Add multiple (similar) rows to definitions table that will be used to add data to an
+#' existing data.table
+#'
+#' @param dtDefs Definition data.table to be modified
+#' @param nvars Number of new variables to define
+#' @param prefix Prefix (character) for new variables
+#' @param formula An R expression for mean (string)
+#' @param variance Number or formula
+#' @param dist Distribution. For possibilities, see details
+#' @param link The link function for the mean, see details
+#' @param id A string indicating the field name for the unique record identifier
+#' @return A data.table named dtName that is an updated data definitions table
+#' @seealso [distributions]
+#' @details The possible data distributions are: `r paste0(.getDists(),collapse = ", ")`.
+#'
+#' @examples
+#' def <- defRepeatAdd(nvars = 4, prefix = "g", formula = "1/3;1/3;1/3", 
+#'     variance = 0, dist = "categorical")
+#' def <- defDataAdd(def, varname = "a", formula = "1;1", dist = "trtAssign")
+#' def <- defRepeatAdd(def, 8, "b", formula = "5 + a", variance = 3, dist = "normal")
+#' def <- defDataAdd(def, "y", formula = "0.10", dist = "binary")
+#' 
```

Comment:
Do we need `defRepeatAdd` in addition to `defRepeat` as they do the same thing? Code duplication like this is usally a big no-no :D

Reply:
Well - they do the same thing, except for the variable checking, just like `defData` and `defDataAdd`. `defRepeatAdd` calls `defDataAdd`, whereas `defRepeat` calls `defData`.

Reply:
Hmm I see. Ah that remindes  me of #18 
But that can still be done without complete duplication of code... I`ll try it later

Reply:
Alternatively, `defRepeat` wouldn't be needed at all if `defData` had another argument `nVars` that would just indicate how many instances of the current variable. But, I don't mind the extra function, since it makes it explicit what the user is trying to do. Again, complete transparency at the expense of efficiency.